drivers: fix and optimization of arc timer

vonhust opened the pull request:
For single core:
   the pulse triggered timer irq doesn't work for all targets. In
   iotdk, we found the clear of IP bit will clear int request
   when elapsed called in thread context. So come back to level
   triggered way, and use the sw triggered irq to remember the
   irq request which may be cleared in thread context with an
   exception that elapsed called in timer irq handler.

For SMP:
* still need to clear IP bit in timer irq handler

* last_time should be aligned to ticks, old code will miss some
  cycles, (curret_time - last_time) % CYC_PER_TICK

* in timeout set, the operations of timer aux regs require some cycles
  , use MIN_DELAY to guarantee this.

Code optimizations:
* rename overflow_cyc to overflow_cycles for better understanding
* use MIN macro to replace if .. else ..

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Fixes #25178 
Fixes #25179

abrodkin commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -90,10 +90,10 @@ static u32_t announced_cycles;
+ * the overflow_cycles must be reset to zero.
  */
-static volatile u32_t overflow_cyc;
+static volatile u32_t overflow_cycles;
```
Cannot we leave this variable rename out of this change? This hides important functional changes in between beautification changes.

vonhust replied:
It's already in different commit

abrodkin commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -384,11 +394,7 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
-		} else {
-			last_load = delay;
-		}
+		last_load = MIN(delay, MAX_CYCLES);
```
I'd suggest to move it to a separate change as well as it has nothing to do with the main reason for this patch - yet another subtle yet quite fundamental redesign of Timer IRQ handling for ARC.

vonhust replied:
It's already in different commit

abrodkin commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -322,7 +331,8 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 
 	/* Desired delay in the future */
-	delay = (ticks == 0) ? CYC_PER_TICK : ticks * CYC_PER_TICK;
+	delay = ticks * CYC_PER_TICK;
```
Same here - it's a good move to use much more readable `MAX` macro but especially given we change a functionality here a little bit (move from `CYC_PER_TICK` to `MIN_DELAY` in one case) I'd prefer to have it as a separate commit with clear explanation on why we can do such a transition. From visible diff context here and existing commit message I cannot understand why that change is possible at all and if it doesn't break anything.

vonhust replied:
already changed

abrodkin reviewed (changes requested):
1. May we have more precise explanation of what and why doesn't work on IoTDK as expected. Why do we see behavior which differs from other boards? Is that ARC EM version that has some (known?) specifics or it's ARC core configuration which leads to such a behavior? In any case I'd like to see a reference either to some documentation or bug-tracker entry which has a valid explanation of this phenomenon.
2. Please de-couple clean-up changes from real meat of this fix/change so we may focus on important changes.
3. Given yet another iteration of our timer driver rework maybe we start from design specification where we may capture known problems and explain how are we trying to get the ultimate implementation that just works? :)

ruuddw commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -190,13 +190,22 @@ static u32_t elapsed(void)
+		 * which may be cleared by the above operation, if called in
+		 * timer irq, no such a need
+		 */
+		if (!z_arc_v2_irq_unit_is_in_isr() ||
```
I think it would be better to not call elapsed() in the timer ISR, but duplicate the relevant part of this code there. Makes things easier to understand and less complex (avoids this check that I can't oversee is always correct).

vonhust replied:
good idea

vonhust replied:
elapsed() may be called multiple in timer_int_handler, one is at the beginning,  others are called in z_clock_announce -> z_clock_set_timeout.

So better to handle it in elapsed()

ruuddw replied:
Understood. Still, I'm not sure this solution is robust, e.g. if another interrupt handler calls elapsed() (reading the timer to get a timestamp for the input data).
And, I'd like to know why the pulse interrupts don't work on IoTDK.

carlescufi commented:
@vonhust will you address the review comments this week? This fixes to high-priority bugs so it's important that we get it merged.

Commit:
drivers: arcv2_timer0: rename overflow_cyc to overflow_cycles

* rename overflow_cyc to overflow_cycles for better understanding
* use MIN macro to replace if .. else ..
* typo fix in comments

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
drivers: arcv2_timer0: back to use level triggered irq

the pulse triggered timer irq doesn't work for all targets. In
iotdk, we found the clear of IP bit will clear int request
when elapsed called in thread context. So come back to level
triggered way which is supported in all targets, and use the sw
triggered irq to remember the irq request which may be cleared
in non timer int handler.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
drivers: arcv2_timer0: minor fix and optimization for SMP case

* still need to clear IP bit in timer irq handler

* last_time should be aligned to ticks, old code will miss some
cycles which are about (curret_time - last_time) % CYC_PER_TICK

* in timeout set, shorten the delay needed when tick is 0, this
 will improve the response of timer irq

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

vonhust commented:
> 1. May we have more precise explanation of what and why doesn't work on IoTDK as expected. Why do we see behavior which differs from other boards? Is that ARC EM version that has some (known?) specifics or it's ARC core configuration which leads to such a behavior? In any case I'd like to see a reference either to some documentation or bug-tracker entry which has a valid explanation of this phenomenon.
> 2. Please de-couple clean-up changes from real meat of this fix/change so we may focus on important changes.
> 3. Given yet another iteration of our timer driver rework maybe we start from design specification where we may capture known problems and explain how are we trying to get the ultimate implementation that just works? :)

I already split this PR into 3 commits.  For the root cause why iotdk doesn't work,  i think  it depends on how it configured in Architect.

vonhust commented:
@ruuddw  @abrodkin ,  now the sanitycheck tests of iotdk are clean.  Let's get it merged to fix the high priority issues, to not block the release of Zephyr.

abrodkin reviewed (approved):
LGTM now, thanks.
